### ISSUE TITLE
AutomaticConfiguration do not work with MultiBeanListProcessor

### ISSUE THREAD

**Issue by rbatista** [uniVocity/univocity-parsers#149]
When I instantiate a ```FixedWidthParser``` with a ```MultiBeanListProcessor``` the ```CommonParserSettings#configureFromAnnotations(beanClass)``` is not called because, it is not an instance of AbstractBeanProcessor. Should not the method be called for each ```AbstractProcessorBean``` in the ```MultiBeanListProcessor```?

The example code:
```
FixedWidthParserSettings settings = new FixedWidthParserSettings();
settings.setAutoConfigurationEnabled(true);
settings.setHeaderExtractionEnabled(false);
settings.getFormat().setLineSeparator("\n");

MultiBeanListProcessor processor = new MultiBeanListProcessor(FileHeader.class, ...);   // FileHeader has an @Headers and fields with @Parsed
settings.setProcessor(processor);

FixedWidthParser parser = new FixedWidthParser(settings);     // Here should call configureFromAnnotations

try (Reader reader = getReader("/positional-file")) {

	parser.parse(reader);   // the exception is throwed here
			
	} catch (IOException e) {
		e.printStackTrace();
	}

```

The exception:
```
com.univocity.parsers.common.DataProcessingException: Could not find fields [bankCode, bankName, batchCode] in input. Please enable header extraction in the parser settings in order to match field names.
Internal state when error was thrown: line=0, column=0, record=1, charIndex=240
	at com.univocity.parsers.common.processor.core.BeanConversionProcessor.mapFieldIndexes(BeanConversionProcessor.java:360)
	at com.univocity.parsers.common.processor.core.BeanConversionProcessor.mapValuesToFields(BeanConversionProcessor.java:289)
	at com.univocity.parsers.common.processor.core.BeanConversionProcessor.createBean(BeanConversionProcessor.java:457)
	at com.univocity.parsers.common.processor.core.AbstractBeanProcessor.rowProcessed(AbstractBeanProcessor.java:51)
	at com.univocity.parsers.common.processor.core.AbstractMultiBeanProcessor.rowProcessed(AbstractMultiBeanProcessor.java:101)
	at com.univocity.parsers.common.Internal.process(Internal.java:21)
	at com.univocity.parsers.common.AbstractParser.rowProcessed(AbstractParser.java:596)
	at com.univocity.parsers.common.AbstractParser.parse(AbstractParser.java:132)
```

The value of ```context.headers()``` at ```BeanConversionProcessor.mapFieldIndexes``` is null.

Is there any other way to use ```MultiBeanListProcessor``` with ```AutoConfiguration``` from ```@Headers```?

**Comment by jbax** [comment]
Can you please add the code of your bean class? Just the attributes with
annotations is enough. Thanks!

**Comment by rbatista** [comment]
This is an abbreviated version of the bean:

```
import com.univocity.parsers.annotations.FixedWidth;
import com.univocity.parsers.annotations.Headers;
import com.univocity.parsers.annotations.Parsed;
import com.univocity.parsers.fixed.FieldAlignment;

@Headers(sequence = { "bankCode", "batchCode", "registerType" }, extract = false, write = false)
public class FileHeader {

	@Parsed
	@FixedWidth(value = 3, alignment = FieldAlignment.RIGHT, padding = '0')
	private Integer bankCode;

	@Parsed
	@FixedWidth(value = 4, alignment = FieldAlignment.RIGHT, padding = '0')
	private Integer batchCode;

	@Parsed
	@FixedWidth(value = 1, alignment = FieldAlignment.RIGHT, padding = '0')
	private Integer registerType;`

     /** getters and setters  */
}
```

**Comment by rbatista** [comment]
Its work if I change the ```MultiBeanListProcessor(FileHeader.class)``` for a ```BeanListProcessor(FileHeader.class)```.

**Comment by jbax** [comment]
Fixed and released a 2.4.2-SNAPSHOT build to handle this.
Thank you!

**Comment by rbatista** [comment]
Thanks @jbax, I tried to do the same thing with the ```InputValueSwitch``` and It has this problem too.

**Comment by rbatista** [comment]
I saw that the ```InputValueSwitch#addSwitchForValue``` can receive an  headers array, but how can I use a different ```FixedWidthFields``` for each processor? I have some processors with different column width. Can you please give me a little help on this?

**Comment by jbax** [comment]
For fixed-width with multiple formats per line you need to define a lookahead and assign field lengths for each possible line, so the parser can know how to collect the values at each row.

You need to get all rows extracted correctly first (without thinking about parsing into beans) and then you can go with the `InputValueSwitch`

**Comment by rbatista** [comment]
I tried something like this:
Processor:
```
StringReader input = new StringReader(
	"34100011C20" +
	"3410001300001" +
	"3410001300002" +
	"34100015         "
);

final FixedWidthFields batchHeaderFields = new FixedWidthFields(BatchHeader.class);
final FixedWidthFields batchDetailFields = new FixedWidthFields(BatchDetailA.class);
final FixedWidthFields batchTailFields = new FixedWidthFields(BatchTrail.class);

FixedWidthParserSettings settings = new FixedWidthParserSettings();
settings.setAutoConfigurationEnabled(true);
settings.setHeaderExtractionEnabled(false);
settings.getFormat().setLineSeparator("\n");

settings.addFormatForLookahead("34100011", batchHeaderFields);
settings.addFormatForLookahead("34100013", batchDetailFields);
settings.addFormatForLookahead("34100015", batchTailFields);
final BeanListProcessor<BatchHeader> batchHeaderProcessor = new BeanListProcessor<BatchHeader>(BatchHeader.class);
final BeanListProcessor<BatchDetailA> batchDetailProcessor = new BeanListProcessor<BatchDetailA>(BatchDetailA.class);
final BeanListProcessor<BatchTrail> batchTailProcessor = new BeanListProcessor<BatchTrail>(BatchTrail.class);

InputValueSwitch inputValueSwitch = new InputValueSwitch(2);
inputValueSwitch.addSwitchForValue("1", batchHeaderProcessor, batchHeaderFields.getFieldNames());
inputValueSwitch.addSwitchForValue("3", batchDetailProcessor, batchDetailFields.getFieldNames());
inputValueSwitch.addSwitchForValue("5", batchTailProcessor, batchTailFields.getFieldNames());

settings.setProcessor(inputValueSwitch);

FixedWidthParser parser = new FixedWidthParser(settings);
parser.parse(input);

List<BatchHeader> headers = batchHeaderProcessor.getBeans();

System.out.println("*** HEADERS ***");
for (BatchHeader header : headers) {
	System.out.println(header);
}

List<BatchDetailA> details = batchDetailProcessor.getBeans();
System.out.println("*** DETAILS ***");
for (BatchDetailA detail : details) {
	System.out.println(detail);
}

System.out.println("*** TAILS ***");
List<BatchTrail> tails = batchTailProcessor.getBeans();
for (BatchTrail tail : tails) {
	System.out.println(tail);
}
```

Beans:
```
public class BatchHeader {

	/** The bank code. */
	@Parsed
	@FixedWidth(value = 3, alignment = FieldAlignment.RIGHT, padding = '0')
	private Integer bankCode;

	/** The batch code. */
	@Parsed
	@FixedWidth(value = 4, alignment = FieldAlignment.RIGHT, padding = '0')
	private Integer batchCode;

	/** The register type. */
	@Parsed
	@FixedWidth(value = 1, alignment = FieldAlignment.RIGHT, padding = '0')
	private Integer registerType;

	@Parsed
	@FixedWidth(value = 1, alignment = FieldAlignment.LEFT, padding = ' ')
	private String operationType;

	@Parsed
	@FixedWidth(value = 2, alignment = FieldAlignment.RIGHT, padding = '0')
	private Integer paymentType;

        /** ... */
}

public class BatchDetailA {
	
	/** The bank code. */
	@Parsed
	@FixedWidth(value = 3, alignment = FieldAlignment.RIGHT, padding = '0')
	private Integer bankCode;

	/** The batch code. */
	@Parsed
	@FixedWidth(value = 4, alignment = FieldAlignment.RIGHT, padding = '0')
	private Integer batchCode;

	/** The register type. */
	@Parsed
	@FixedWidth(value = 1, alignment = FieldAlignment.RIGHT, padding = '0')
	private Integer registerType;
	
	@Parsed
	@FixedWidth(value = 5, alignment = FieldAlignment.RIGHT, padding = '0')
	private Integer registerNumber;

        /** ... */
}

	/** The bank code. */
	@Parsed
	@FixedWidth(value = 3, alignment = FieldAlignment.RIGHT, padding = '0')
	private Integer bankCode;

	/** The batch code. */
	@Parsed
	@FixedWidth(value = 4, alignment = FieldAlignment.RIGHT, padding = '0')
	private Integer batchCode;

	/** The register type. */
	@Parsed
	@FixedWidth(value = 1, alignment = FieldAlignment.RIGHT, padding = '0')
	private Integer registerType;
	
	@Parsed
	@FixedWidth(value = 9, alignment = FieldAlignment.LEFT, padding = ' ')
	private Integer blank;

       /** ... */
}
```

And the output was like I expected:
```
*** HEADERS ***
BatchHeader [bankCode=341, batchCode=1, registerType=1, operationType=C, paymentType=20]
*** DETAILS ***
BatchDetailA [bankCode=341, batchCode=1, registerType=3, registerNumber=1]
BatchDetailA [bankCode=341, batchCode=1, registerType=3, registerNumber=2]
*** TAILS ***
BatchTrail [bankCode=341, batchCode=1, registerType=5, blank=null]
PASSED: testBatches
```

Thank you.

Ps.: There is some wildcard to be used in the lookahead? I need to parse the row based on the registerType (8th character) and bypass the first 7 characters.

**Comment by jbax** [comment]
Grab 2.5.0-SNAPSHOT and it will allow you to define a wildcard for the lookahead/behind.

For example: `settings.addFormatForLookahead("???????1", batchHeaderFields);`

Abracos e muito obrigado pela sugestao.

**Comment by rbatista** [comment]
Nice! :clap: 
Thank you for all your help.

Another question about lookahead, but now in the `FixedWithWriter`:
Like I saw, I need match the 8th character, and my bean has three fields to match until the 8th char:[bankCode=341, batchCode=0001, registerType=1, ...]

I saw in the `FixedWidthWriter.processRow()` in line [172](https://github.com/uniVocity/univocity-parsers/blob/master/src/main/java/com/univocity/parsers/fixed/FixedWidthWriter.java#L172) the following code:
```
    String value = String.valueOf(row[0]);
    int end;
    if (value.length() >= lookupChars.length) {
        end = lookupChars.length;
    } else {
        end = value.length();
    for (int i = lookupChars.length - 1; i > end; i--) {
        lookupChars[i] = '\0';                     // EMPHASIS HERE
    }
}
```
How I'm trying to match more than one field this does not work because '\0' will never match with the remainder of the lookaheadFormats (line 178):
```
if (lookaheadFormats[i].matches(lookupChars)) {
```

So, you think that is possible define the column to match on the lookahead (in line 172) or replace the '\0' for the next fields until the size of the lookahead?

I would like to thank you again!

Abraço
